File: android/app/src/main/java/com/oasismall/oasisai/util/PngMetadata.kt
package com.oasismall.oasisai.util

import java.io.File
import java.io.RandomAccessFile
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.zip.CRC32

/**
 * Reads/writes barcode in PNG text chunks (visible in file Properties → Details on Windows).
 * - [KEY_BARCODE] — machine-readable for the app
 * - [KEY_DESCRIPTION] — human-readable file description
 */
object PngMetadata {
    const val KEY_BARCODE = "Barcode"
    const val KEY_DESCRIPTION = "Description"
    const val KEY_DESIGNATION = "Designation"
    const val KEY_PRICE_NOW = "PriceNow"
    const val KEY_PRICE_BEFORE = "PriceBefore"
    const val KEY_RAYON = "Rayon"
    const val KEY_CODEART = "Codeart"

    private val PNG_SIGNATURE = byteArrayOf(
        0x89.toByte(), 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A,
    )
    private val BARCODE_FILENAME_SUFFIX = Regex("_([0-9]{8,18})$")
    private val BARCODE_FILENAME_ONLY = Regex("^[0-9]{8,18}$")
    private val BARCODE_IN_DESCRIPTION = Regex("""(?i)barcode\s*[=:]\s*([0-9]{8,18})""")

    data class PngArticleDetails(
        val barcode: String? = null,
        val codeart: String? = null,
        val designation: String? = null,
        val priceNow: Double? = null,
        val priceBefore: Double? = null,
        val rayon: String? = null,
        val description: String? = null,
    )


    fun readAllTextChunks(file: File): Map<String, String> {
        if (!file.isFile || file.extension.lowercase() != "png") return emptyMap()
        val out = mutableMapOf<String, String>()
        RandomAccessFile(file, "r").use { raf ->
            val sig = ByteArray(8)
            if (raf.read(sig) != 8 || !sig.contentEquals(PNG_SIGNATURE)) return emptyMap()
            while (raf.filePointer < raf.length()) {
                val lenBuf = ByteArray(4)
                if (raf.read(lenBuf) != 4) break
                val length = ByteBuffer.wrap(lenBuf).order(ByteOrder.BIG_ENDIAN).int
                if (length < 0 || length > 1_000_000) break
                val type = ByteArray(4)
                if (raf.read(type) != 4) break
                val chunkType = String(type, Charsets.US_ASCII)
                val data = ByteArray(length)
                if (length > 0 && raf.read(data) != length) break
                raf.skipBytes(4)
                if (chunkType == "tEXt" || chunkType == "iTXt") {
                    parseTextChunk(data, chunkType)?.let { (kw, text) ->
                        out[kw] = text
                    }
                }
                if (chunkType == "IEND") break
            }
        }
        return out
    }

    fun readArticleDetails(file: File): PngArticleDetails {
        val chunks = readAllTextChunks(file)
        val barcode = chunks[KEY_BARCODE]?.trim()?.takeIf { it.isNotBlank() }
            ?: chunks[KEY_DESCRIPTION]?.let { desc ->
                BARCODE_IN_DESCRIPTION.find(desc)?.groupValues?.get(1)
            }
            ?: extractBarcodeFromFilename(file.nameWithoutExtension)
        return PngArticleDetails(
            barcode = barcode,
            codeart = chunks[KEY_CODEART]?.trim()?.takeIf { it.isNotBlank() },
            designation = chunks[KEY_DESIGNATION]?.trim()?.takeIf { it.isNotBlank() },
            priceNow = chunks[KEY_PRICE_NOW]?.let { parsePriceText(it) },
            priceBefore = chunks[KEY_PRICE_BEFORE]?.let { parsePriceText(it) },
            rayon = chunks[KEY_RAYON]?.trim()?.takeIf { it.isNotBlank() },
            description = chunks[KEY_DESCRIPTION]?.trim()?.takeIf { it.isNotBlank() },
        )
    }

    /** Parses values like `90,00 DA` from PNG tEXt (matches PC script + [PriceFormatter]). */
    fun parsePriceText(raw: String): Double? {
        val cleaned = raw.trim()
            .removeSuffix("DA")
            .trim()
            .replace(" ", "")
            .replace(",", ".")
        return cleaned.toDoubleOrNull()
    }

    fun readBarcode(file: File): String? =
        readArticleDetails(file).barcode

    /** Safe filename stem for barcode-only PNGs (matches [extractBarcodeFromFilename] rules). */
    fun barcodeFileStem(barcode: String): String {
        val trimmed = barcode.trim()
        if (BARCODE_FILENAME_ONLY.matches(trimmed)) return trimmed
        return trimmed.replace(Regex("[^A-Za-z0-9]"), "").take(32).ifBlank { "barcode" }
    }

    fun buildDescription(barcode: String, designation: String?): String {
        val des = designation?.trim().orEmpty()
        return if (des.isNotEmpty()) {
            "Barcode: $barcode | $des"
        } else {
            "Barcode: $barcode"
        }
    }

    fun writeBarcode(file: File, barcode: String, designation: String?) {
        if (!file.isFile) return
        val description = buildDescription(barcode, designation)
        writeTextChunks(
            file,
            mapOf(
                KEY_BARCODE to barcode.trim(),
                KEY_DESCRIPTION to description,
            ),
        )
    }

    fun writeArticleDetails(
        file: File,
        barcode: String,
        designation: String?,
        priceNow: Double?,
        priceBefore: Double?,
        rayon: String?,
        codeart: String? = null,
    ) {
        if (!file.isFile) return
        val details = buildList {
            designation?.trim()?.takeIf { it.isNotBlank() }?.let { add("Designation: $it") }
            priceNow?.let { add("Price now: ${PriceFormatter.format(it)}") }
            priceBefore?.let { add("Price before: ${PriceFormatter.format(it)}") }
            add("Barcode: ${barcode.trim()}")
            codeart?.trim()?.takeIf { it.isNotBlank() }?.let { add("Code: $it") }
            rayon?.trim()?.takeIf { it.isNotBlank() }?.let { add("Rayon: $it") }
        }
        val chunks = buildMap {
            put(KEY_BARCODE, barcode.trim())
            put(KEY_DESCRIPTION, details.joinToString(" | "))
            designation?.trim()?.takeIf { it.isNotBlank() }?.let { put(KEY_DESIGNATION, it) }
            priceNow?.let { put(KEY_PRICE_NOW, PriceFormatter.format(it)) }
            priceBefore?.let { put(KEY_PRICE_BEFORE, PriceFormatter.format(it)) }
            codeart?.trim()?.takeIf { it.isNotBlank() }?.let { put(KEY_CODEART, it) }
            rayon?.trim()?.takeIf { it.isNotBlank() }?.let { put(KEY_RAYON, it) }
        }
        writeTextChunks(file, chunks)
    }

    fun extractBarcodeFromFilename(stem: String): String? =
        if (BARCODE_FILENAME_ONLY.matches(stem)) {
            stem
        } else {
            BARCODE_FILENAME_SUFFIX.find(stem)?.groupValues?.get(1)
        }

    fun stemWithoutBarcodeSuffix(stem: String): String =
        stem.replace(BARCODE_FILENAME_SUFFIX, "")

    private fun readTextChunk(file: File, keyword: String): String? {
        RandomAccessFile(file, "r").use { raf ->
            val sig = ByteArray(8)
            if (raf.read(sig) != 8 || !sig.contentEquals(PNG_SIGNATURE)) return null
            while (raf.filePointer < raf.length()) {
                val lenBuf = ByteArray(4)
                if (raf.read(lenBuf) != 4) break
                val length = ByteBuffer.wrap(lenBuf).order(ByteOrder.BIG_ENDIAN).int
                if (length < 0 || length > 1_000_000) break
                val type = ByteArray(4)
                if (raf.read(type) != 4) break
                val chunkType = String(type, Charsets.US_ASCII)
                val data = ByteArray(length)
                if (length > 0 && raf.read(data) != length) break
                raf.skipBytes(4) // CRC
                if (chunkType == "tEXt" || chunkType == "iTXt") {
                    parseTextChunk(data, chunkType)?.let { (kw, text) ->
                        if (kw == keyword) return text
                    }
                }
                if (chunkType == "IEND") break
            }
        }
        return null
    }

    private fun parseTextChunk(data: ByteArray, type: String): Pair<String, String>? {
        if (type == "tEXt") {
            val zero = data.indexOf(0)
            if (zero < 0) return null
            val keyword = String(data, 0, zero, Charsets.US_ASCII)
            val text = String(data, zero + 1, data.size - zero - 1, Charsets.ISO_8859_1)
            return keyword to text
        }
        // iTXt: keyword\0 compression\0 method\0 language\0 translated\0 text
        val zero = data.indexOf(0)
        if (zero < 0) return null
        val keyword = String(data, 0, zero, Charsets.US_ASCII)
        if (zero + 1 >= data.size) return null
        val compression = data[zero + 1]
        var pos = zero + 2
        while (pos < data.size && data[pos] != 0.toByte()) pos++
        if (pos >= data.size) return null
        pos++
        while (pos < data.size && data[pos] != 0.toByte()) pos++
        if (pos >= data.size) return null
        pos++
        while (pos < data.size && data[pos] != 0.toByte()) pos++
        if (pos >= data.size) return null
        pos++
        while (pos < data.size && data[pos] != 0.toByte()) pos++
        if (pos >= data.size) return null
        pos++
        if (pos > data.size) return null
        val textBytes = data.copyOfRange(pos, data.size)
        val text = if (compression.toInt() == 0) {
            String(textBytes, Charsets.UTF_8)
        } else {
            return null
        }
        return keyword to text
    }

    private fun writeTextChunks(file: File, chunks: Map<String, String>) {
        val original = file.readBytes()
        if (original.size < 12 || !original.copyOfRange(0, 8).contentEquals(PNG_SIGNATURE)) return
        val rebuilt = stripTextChunks(original, chunks.keys)
        val iendIndex = findIendIndex(rebuilt) ?: return
        val before = rebuilt.copyOfRange(0, iendIndex)
        val after = rebuilt.copyOfRange(iendIndex, rebuilt.size)
        val newChunks = chunks.map { (keyword, text) -> buildTextChunk(keyword, text) }
        val combined = before + newChunks.fold(ByteArray(0)) { acc, c -> acc + c } + after
        file.writeBytes(combined)
    }

    private fun stripTextChunks(png: ByteArray, keywords: Set<String>): ByteArray {
        val out = java.io.ByteArrayOutputStream()
        var offset = 8
        out.write(png, 0, 8)
        while (offset + 12 <= png.size) {
            val length = ByteBuffer.wrap(png, offset, 4).order(ByteOrder.BIG_ENDIAN).int
            if (length < 0) break
            val type = String(png, offset + 4, 4, Charsets.US_ASCII)
            val total = 12 + length
            if (offset + total > png.size) break
            val drop = if (type == "tEXt" || type == "iTXt") {
                val data = png.copyOfRange(offset + 8, offset + 8 + length)
                val parsed = parseTextChunk(data, type)
                parsed != null && parsed.first in keywords
            } else {
                false
            }
            if (!drop) {
                out.write(png, offset, total)
            }
            offset += total
            if (type == "IEND") break
        }
        return out.toByteArray()
    }

    private fun findIendIndex(png: ByteArray): Int? {
        var offset = 8
        while (offset + 12 <= png.size) {
            val length = ByteBuffer.wrap(png, offset, 4).order(ByteOrder.BIG_ENDIAN).int
            if (length < 0 || offset + 12 + length > png.size) return null
            val type = String(png, offset + 4, 4, Charsets.US_ASCII)
            if (type == "IEND") return offset
            offset += 12 + length
        }
        return null
    }

    private fun buildTextChunk(keyword: String, text: String): ByteArray {
        val data = keyword.toByteArray(Charsets.US_ASCII) +
            byteArrayOf(0) +
            text.toByteArray(Charsets.ISO_8859_1)
        val type = "tEXt".toByteArray(Charsets.US_ASCII)
        val length = ByteBuffer.allocate(4).order(ByteOrder.BIG_ENDIAN).putInt(data.size).array()
        val crcInput = type + data
        val crc = CRC32().apply { update(crcInput) }.value.toInt()
        val crcBytes = ByteBuffer.allocate(4).order(ByteOrder.BIG_ENDIAN).putInt(crc).array()
        return length + type + data + crcBytes
    }
}

File: scripts/png_metadata.py
"""Embed/read barcode in PNG tEXt chunks (Windows file Properties → Description)."""

from __future__ import annotations

import re
import struct
import zlib
from pathlib import Path

PNG_SIG = b"\x89PNG\r\n\x1a\n"
KEY_BARCODE = b"Barcode"
KEY_DESCRIPTION = b"Description"
KEY_DESIGNATION = b"Designation"
KEY_PRICE_NOW = b"PriceNow"
KEY_PRICE_BEFORE = b"PriceBefore"
KEY_RAYON = b"Rayon"
KEY_CODEART = b"Codeart"

ARTICLE_TEXT_KEYS = {
    KEY_BARCODE,
    KEY_DESCRIPTION,
    KEY_DESIGNATION,
    KEY_PRICE_NOW,
    KEY_PRICE_BEFORE,
    KEY_RAYON,
    KEY_CODEART,
}
BARCODE_FILENAME_SUFFIX = re.compile(r"_(\d{8,18})$")
BARCODE_FILENAME_ONLY = re.compile(r"^\d{8,18}$")
BARCODE_IN_DESCRIPTION = re.compile(r"(?i)barcode\s*[=:]\s*([0-9]{8,18})")


def format_price_da(price: float) -> str:
    """Match Android PriceFormatter (French grouping + DA)."""
    negative = price < 0
    price = abs(price)
    whole = int(price)
    cents = int(round((price - whole) * 100))
    if cents == 100:
        whole += 1
        cents = 0
    s = str(whole)
    groups: list[str] = []
    while len(s) > 3:
        groups.insert(0, s[-3:])
        s = s[:-3]
    groups.insert(0, s)
    int_part = " ".join(groups)
    body = f"{int_part},{cents:02d} DA"
    return f"-{body}" if negative else body


def build_description(barcode: str, designation: str | None) -> str:
    des = (designation or "").strip()
    if des:
        return f"Barcode: {barcode} | {des}"
    return f"Barcode: {barcode}"


def read_barcode_from_chunks(path: Path) -> str | None:
    """Barcode from PNG tEXt only (not filename)."""
    data = path.read_bytes()
    if not data.startswith(PNG_SIG):
        return None
    offset = 8
    while offset + 12 <= len(data):
        length = struct.unpack(">I", data[offset : offset + 4])[0]
        chunk_type = data[offset + 4 : offset + 8]
        chunk_data = data[offset + 8 : offset + 8 + length]
        offset += 12 + length
        if chunk_type == b"tEXt":
            zero = chunk_data.find(b"\x00")
            if zero >= 0:
                keyword = chunk_data[:zero]
                text = chunk_data[zero + 1 :].decode("latin-1", errors="replace")
                if keyword == KEY_BARCODE and text.strip():
                    return text.strip()
                if keyword == KEY_DESCRIPTION:
                    m = BARCODE_IN_DESCRIPTION.search(text)
                    if m:
                        return m.group(1)
        if chunk_type == b"IEND":
            break
    return None


def read_barcode(path: Path) -> str | None:
    return read_barcode_from_chunks(path) or _barcode_from_filename(path.stem)


def _barcode_from_filename(stem: str) -> str | None:
    if BARCODE_FILENAME_ONLY.fullmatch(stem):
        return stem
    m = BARCODE_FILENAME_SUFFIX.search(stem)
    return m.group(1) if m else None


def _crc(chunk_type: bytes, chunk_data: bytes) -> bytes:
    return struct.pack(">I", zlib.crc32(chunk_type + chunk_data) & 0xFFFFFFFF)


def _text_chunk(keyword: str, text: str) -> bytes:
    chunk_data = keyword.encode("ascii") + b"\x00" + text.encode("latin-1", errors="replace")
    chunk_type = b"tEXt"
    return (
        struct.pack(">I", len(chunk_data))
        + chunk_type
        + chunk_data
        + _crc(chunk_type, chunk_data)
    )


def _strip_keywords(png: bytes, keywords: set[bytes]) -> bytes:
    if not png.startswith(PNG_SIG):
        return png
    out = bytearray(PNG_SIG)
    offset = 8
    while offset + 12 <= len(png):
        length = struct.unpack(">I", png[offset : offset + 4])[0]
        chunk_type = png[offset + 4 : offset + 8]
        total = 12 + length
        chunk = png[offset : offset + total]
        drop = False
        if chunk_type in (b"tEXt", b"iTXt"):
            chunk_data = png[offset + 8 : offset + 8 + length]
            if chunk_type == b"tEXt":
                zero = chunk_data.find(b"\x00")
                if zero >= 0:
                    kw = chunk_data[:zero]
                    if kw in keywords:
                        drop = True
        if not drop:
            out.extend(chunk)
        offset += total
        if chunk_type == b"IEND":
            break
    return bytes(out)


def _find_iend_offset(png: bytes) -> int | None:
    """Byte offset of the real IEND chunk (never use rfind — IEND can appear inside IDAT)."""
    offset = 8
    while offset + 12 <= len(png):
        length = struct.unpack(">I", png[offset : offset + 4])[0]
        chunk_type = png[offset + 4 : offset + 8]
        if chunk_type == b"IEND":
            return offset
        if length < 0 or length > 50_000_000:
            return None
        offset += 12 + length
    return None


def build_article_description(
    *,
    barcode: str,
    designation: str | None,
    codeart: str | None,
    price_now: float | None,
    price_before: float | None = None,
    rayon: str | None,
) -> str:
    parts: list[str] = []
    if designation and designation.strip():
        parts.append(f"Designation: {designation.strip()}")
    if price_now is not None:
        parts.append(f"Price now: {format_price_da(price_now)}")
    if price_before is not None:
        parts.append(f"Price before: {format_price_da(price_before)}")
    parts.append(f"Barcode: {barcode.strip()}")
    if codeart and codeart.strip():
        parts.append(f"Code: {codeart.strip()}")
    if rayon and rayon.strip():
        parts.append(f"Rayon: {rayon.strip()}")
    return " | ".join(parts)


def write_article_details(
    path: Path,
    *,
    barcode: str,
    designation: str | None,
    codeart: str | None,
    price_now: float | None,
    price_before: float | None = None,
    rayon: str | None,
) -> None:
    """Write tEXt chunks aligned with Android PngMetadata."""
    png = path.read_bytes()
    if not png.startswith(PNG_SIG):
        return
    stripped = _strip_keywords(png, ARTICLE_TEXT_KEYS)
    iend = _find_iend_offset(stripped)
    if iend is None:
        return

    description = build_article_description(
        barcode=barcode,
        designation=designation,
        codeart=codeart,
        price_now=price_now,
        price_before=price_before,
        rayon=rayon,
    )
    chunks = [_text_chunk("Barcode", barcode.strip()), _text_chunk("Description", description)]
    if designation and designation.strip():
        chunks.append(_text_chunk("Designation", designation.strip()))
    if codeart and codeart.strip():
        chunks.append(_text_chunk("Codeart", codeart.strip()))
    if price_now is not None:
        chunks.append(_text_chunk("PriceNow", format_price_da(price_now)))
    if price_before is not None:
        chunks.append(_text_chunk("PriceBefore", format_price_da(price_before)))
    if rayon and rayon.strip():
        chunks.append(_text_chunk("Rayon", rayon.strip()))

    updated = stripped[:iend] + b"".join(chunks) + stripped[iend:]
    path.write_bytes(updated)


def write_barcode(path: Path, barcode: str, designation: str | None) -> None:
    png = path.read_bytes()
    if not png.startswith(PNG_SIG):
        return
    stripped = _strip_keywords(png, {KEY_BARCODE, KEY_DESCRIPTION})
    iend = _find_iend_offset(stripped)
    if iend is None:
        return
    description = build_description(barcode, designation)
    new_chunks = _text_chunk("Barcode", barcode.strip()) + _text_chunk(
        "Description", description
    )
    updated = stripped[:iend] + new_chunks + stripped[iend:]
    path.write_bytes(updated)
